cluster/state/mutationaddvalidator.go

```diff
@@ -33,6 +41,85 @@ func (v addValidators) MarshalJSON() ([]byte, error) {
 	return b, nil
 }
 
+// Verify validates the gen validators mutation, ensuring validators are populated with valid addresses.
+// This allows ToSSZ to be called without error.
+func (v genValidators) Verify() error {
+	if len(v.Validators) == 0 {
+		return errors.New("no validators")
+	}
+
+	for _, validator := range v.Validators {
+		if _, err := from0xHex(validator.FeeRecipientAddress, 20); err != nil {
+			return errors.Wrap(err, "validate fee recipient address")
+		}
+		if _, err := from0xHex(validator.WithdrawalAddress, 20); err != nil {
+			return errors.Wrap(err, "validate withdrawal address")
+		}
+	}
+
+	return nil
+}
+
+func transformGenValidators(c Cluster, signed SignedMutation) (Cluster, error) {
+	if err := verifyEmptySig(signed); err != nil {
+		return c, errors.Wrap(err, "verify empty sig")
+	}
+
+	if signed.Mutation.Type != TypeGenValidators {
+		return c, errors.New("invalid mutation type")
+	}
+
+	gen, ok := signed.Mutation.Data.(genValidators)
+	if !ok {
+		return c, errors.New("invalid gen validators data")
+	}
+
+	c.Validators = append(c.Validators, gen.Validators...)
+
+	return c, nil
+}
+
+// NewAddValidators creates a new composite add validators mutation from the provided gen validators and node approvals.
+func NewAddValidators(genValidators, nodeApprovals SignedMutation) (SignedMutation, error) {
+	if genValidators.Mutation.Type != TypeGenValidators {
+		return SignedMutation{}, errors.New("invalid gen validators mutation type")
+	}
+
+	if nodeApprovals.Mutation.Type != TypeNodeApprovals {
+		return SignedMutation{}, errors.New("invalid node approvals mutation type")
+	}
+
+	return SignedMutation{
+		Mutation: Mutation{
+			Parent:    genValidators.Mutation.Parent,
+			Type:      TypeAddValidators,
+			Timestamp: nowFunc(),
+			Data:      addValidators{GenValidators: genValidators, NodeApprovals: nodeApprovals},
+		},
+		// Composite mutations have no signer or signature.
+	}, nil
+}
+
+type addValidators struct {
+	GenValidators SignedMutation `ssz:"Composite" json:"gen_validators"`
+	NodeApprovals SignedMutation `ssz:"Composite" json:"node_approvals"`
+}
+
+func (a addValidators) MarshalJSON() ([]byte, error) {
+	b, err := json.Marshal(struct {
```

Comment:
nit: this struct can be `addValidatorsJSON`

Reply:
cool